File: Cpp/generate_parenthesis.cpp
class Solution {
public:
    void backtrack(vector<string> & list_par, int open_count, int close_count, int max_count, string stack_par)
    {
        if (open_count == max_count && close_count == max_count)
        {
            list_par.push_back(stack_par);
            return;
        }
        if (open_count < max_count) backtrack(list_par, open_count+1, close_count, max_count, stack_par + "(");
        if (open_count > close_count) backtrack(list_par, open_count, close_count+1, max_count, stack_par + ")");
    }


    vector<string> generateParenthesis(int n) {
        vector<string> list_par;
        string stack_par;
        backtrack(list_par, 0, 0, n, stack_par);
        return list_par;
    }


};

File: Python/generate_parenthesis.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        list_par = []
        stack = []
        def backtrack(open_count, close_count):
            if open_count == close_count == n:
                list_par.append("".join(stack))
                return

            if  close_count < open_count:
                stack.append(')')
                backtrack(open_count, close_count+1)
                stack.pop()
            
            if  open_count < n:
                stack.append('(')
                backtrack(open_count+1, close_count)
                stack.pop()
        backtrack(0,0)
        return list_par